README.md
# d3v4_usage_example
_Tests for D3 usages_

_Created by alicezou_


The official website for d3 is https://d3js.org/.
You can find many examples on https://observablehq.com/@d3/gallery.

However, I still find D3 has limited documentations and needs lots of trials and errors.

The examples in this repo is for D3v4.

Many examples from the internet are for d3v3. D3v4 and D3v3 have many differences, and you cannot simply copy and paste to use code from v3 to v4.


The name for testing javascript and testing html are the same.

Have fun. :)

CHANGES.md
# Changes in D3 6.0

[Released August 26, 2020.](https://github.com/d3/d3/releases/tag/v6.0.0)

*This document covers only major changes. For minor and patch changes, please see the [release notes](https://github.com/d3/d3/releases).*

D3 now **uses native collections** (Map and Set) and **accepts iterables**. [d3.group and d3.rollup](https://observablehq.com/@d3/d3-group) are powerful new aggregation functions that replace d3.nest and work great [with d3-hierarchy](https://observablehq.com/d/9a453665f405eebf) and d3-selection. There are lots of new helpers in d3-array, too, such as [d3.greatest](https://observablehq.com/@d3/d3-least), [d3.quickselect](https://observablehq.com/@d3/d3-quickselect), and [d3.fsum](https://observablehq.com/@d3/d3-fsum).

D3 now **passes events directly to listeners**, replacing the d3.event global and bringing D3 inline with vanilla JavaScript and most other frameworks.

**d3-delaunay** (based on Vladimir Agafonkin’s excellent [Delaunator](https://github.com/mapbox/delaunator)) replaces d3-voronoi, offering dramatic improvements to performance, robustness, and [search](https://observablehq.com/@d3/delaunay-find). And there’s a new [d3-geo-voronoi](https://github.com/Fil/d3-geo-voronoi) for spherical (geographical) data! **d3-random** is [greatly expanded](https://github.com/d3/d3-random/blob/master/README.md) and includes a fast [linear congruential generator](https://observablehq.com/@d3/d3-randomlcg) for seeded randomness. **d3-chord** has new layouts for [directed](https://observablehq.com/@d3/directed-chord-diagram) and transposed chord diagrams. **d3-scale** adds a new [radial scale](https://observablehq.com/@d3/radial-stacked-bar-chart-ii) type.

… and a variety of other small enhancements. [More than 450 examples](https://observablehq.com/@d3/gallery) have been updated to D3 6.0!

### d3-array

* Accept iterables.
* Add [d3.group](https://github.com/d3/d3-array/blob/master/README.md#group).
* Add [d3.groups](https://github.com/d3/d3-array/blob/master/README.md#groups).
* Add [d3.index](https://github.com/d3/d3-array/blob/master/README.md#index).
* Add [d3.indexes](https://github.com/d3/d3-array/blob/master/README.md#indexes).
* Add [d3.rollup](https://github.com/d3/d3-array/blob/master/README.md#rollup).
* Add [d3.rollups](https://github.com/d3/d3-array/blob/master/README.md#rollups).
* Add [d3.maxIndex](https://github.com/d3/d3-array/blob/master/README.md#maxIndex).
* Add [d3.minIndex](https://github.com/d3/d3-array/blob/master/README.md#minIndex).
* Add [d3.greatest](https://github.com/d3/d3-array/blob/master/README.md#greatest).
* Add [d3.greatestIndex](https://github.com/d3/d3-array/blob/master/README.md#greatestIndex).
* Add [d3.least](https://github.com/d3/d3-array/blob/master/README.md#least).
* Add [d3.leastIndex](https://github.com/d3/d3-array/blob/master/README.md#leastIndex).
* Add [d3.bin](https://github.com/d3/d3-array/blob/master/README.md#bin).
* Add [d3.count](https://github.com/d3/d3-array/blob/master/README.md#count).
* Add [d3.cumsum](https://github.com/d3/d3-array/blob/master/README.md#cumsum).
* Add [d3.fsum](https://github.com/d3/d3-array/blob/master/README.md#fsum).
* Add [d3.Adder](https://github.com/d3/d3-array/blob/master/README.md#Adder).
* Add [d3.quantileSorted](https://github.com/d3/d3-array/blob/master/README.md#quantileSorted).
* Add [d3.quickselect](https://github.com/d3/d3-array/blob/master/README.md#quickselect).
* Add [*bisector*.center](https://github.com/d3/d3-array/blob/master/README.md#bisector_center).
* Allow more than two iterables for [d3.cross](https://github.com/d3/d3-array/blob/master/README.md#cross).
* Accept non-sorted input with [d3.quantile](https://github.com/d3/d3-array/blob/master/README.md#quantile).
* Fix a *array*.sort bug in Safari.
* Fix bin thresholds to ignore NaN input.
* Fix [d3.ticks](https://github.com/d3/d3-array/blob/master/README.md#ticks) to not return ticks outside the domain.
* Improve the performance of [d3.median](https://github.com/d3/d3-array/blob/master/README.md#median).

See https://observablehq.com/@d3/d3-array-2-0 for details.

### d3-brush

* Add [*event*.mode](https://github.com/d3/d3-brush/blob/master/README.md#brush-events).
* Change [*brush*.on](https://github.com/d3/d3-brush/blob/master/README.md#brush_on) to pass the *event* directly to listeners.
* Improve multitouch (two-touch) interaction.

### d3-chord

* Add [d3.chordDirected](https://github.com/d3/d3-chord/blob/master/README.md#chordDirected).
* Add [d3.chordTranspose](https://github.com/d3/d3-chord/blob/master/README.md#chordTranspose).
* Add [d3.ribbonArrow](https://github.com/d3/d3-chord/blob/master/README.md#ribbonArrow).
* Add [*ribbon*.padAngle](https://github.com/d3/d3-chord/blob/master/README.md#ribbon_padAngle).
* Add [*ribbon*.sourceRadius](https://github.com/d3/d3-chord/blob/master/README.md#ribbon_sourceRadius).
* Add [*ribbon*.targetRadius](https://github.com/d3/d3-chord/blob/master/README.md#ribbon_targetRadius).

### d3-delaunay

* Add [d3.Delaunay](https://github.com/d3/d3-delaunay/blob/master/README.md).

### d3-drag

* Change [*drag*.on](https://github.com/d3/d3-drag/blob/master/README.md#drag_on) to pass the *event* directly to listeners.

### d3-force

* Add *iterations* argument to [*simulation*.tick](https://github.com/d3/d3-force/blob/master/README.md#simulation_tick).
* Add [*forceCenter*.strength](https://github.com/d3/d3-force/blob/master/README.md#center_strength).
* Add [*forceSimulation*.randomSource](https://github.com/d3/d3-force/blob/master/README.md#simulation_randomSource).
* All built-in forces are now fully deterministic (including “jiggling” coincident nodes).
* Improve the default phyllotaxis layout slightly by offsetting by one half-radius.
* Improve the error message when a link references an unknown node.
* [*force*.initialize](https://github.com/d3/d3-force/blob/master/README.md#force_initialize) is now passed a random source.

test.js
var svg = d3.select("svg");

var width = svg.attr("width");
var height = svg.attr("height");
var url_pre = "localhost:5000/";
svg = svg.call(d3.zoom().on("zoom", zoomed)).append("g");
/*svg.append("defs").append("marker")
    .attr("id", "arrow")
    .attr("viewBox", "0 -5 10 10")
    .attr("refX", 20)
    .attr("refY", 0)
    .attr("markerWidth", 8)
    .attr("markerHeight", 8)
    .attr("orient", "auto")
    .append("svg:path")
    .attr("d", "M0,-5L10,0L0,5");*/
var color = d3.scaleOrdinal(d3.schemeCategory10);

var dummy = [0];
    var retangulo = svg.append("rect")
    		.attrs({
          "x": 480,
          "y": 141,
          "width": 40,
          "height": 30
        })
    		.style("fill", "blue");
svg.append("g").append("rect")
    .attrs({
        "x": 480,
        "y": 141,
        "width": 40,
        "height": 30
    })
    .style("fill", "blue");

/*
var legend_line = svg.append("g").attr("class", "legend_line").selectAll("legend_line")
    .enter().append("circle")
    .data(dummy)
    .attr("cx", -100)
    .attr("cy", 100)
    .attr("r", 20)
    .style("fill", color(1));
*/


function dragstarted(d) {
    if (!d3.event.active) simulation.alphaTarget(0.3).restart();
    d.fx = d.x;
    d.fy = d.y;
}

function dragged(d) {
    d.fx = d3.event.x;
    d.fy = d3.event.y;
}

function dragended(d) {
    if (!d3.event.active) simulation.alphaTarget(0);
    d.fx = null;
    d.fy = null;
}

function zoomed() {
    svg.attr("transform", "translate(" + d3.event.transform.x + "," + d3.event.transform.y + ")" + " scale(" + d3.event.transform.k + ")");
}

function openLink() {
    return function (d) {
        var url = "";
        if (d.id != "") {
            url = url_pre.concat(d.id)
        } else {
            console.log("ERROR: No url")
        }
        window.open(url)
    }
}


/*
var data = {
    "val": [
        {
            "id": "val/0",
            "type": 1,
            "creater":"UserA"
        },
        {
            "id": "val/1",
            "type": 1,
            "creater":"UserA"
        }],
    "var": [{
        "id": "var/0",
        "type": 2,
        "creater":"UserB"
    }],
    "links": [
        {
            "source": "var/0",
            "target": "val/1",
            "type": 1
        }]
};*/

// createGraph(false, data);



test_search.js
var width = 900,
    height = 590;
    
var zoom = d3.behavior.zoom().scaleExtent([0.1,5]).on("zoom", redraw);
var svg = d3.select("body")
    .append("svg")
    .attr("width", width)
    .attr("height", height)
    .call(zoom)
    .on("dblclick.zoom", null)
    .append('g');

//INPUT DATA
var links = [
    {source: 'N1', target: 'N2'},
    {source: 'N1', target: 'N3'},
    {source: 'N2', target: 'N3'},
    {source: 'N3', target: 'N4'},
];
  
var nodes = [ 
    {id: 'N1', name: 'A'},
    {id: 'N2', name: 'B'},
    {id: 'N3', name: 'C'},
    {id: 'N4', name: 'D'},
];

//CONNECTIONS
var hash_lookup = [];
nodes.forEach(function(d, i) {
  hash_lookup[d.id] = d;
});
links.forEach(function(d, i) {
  d.source = hash_lookup[d.source];
  d.target = hash_lookup[d.target];
});

//FORCE LAYOUT
var force = d3.layout.force()
    .size([width, height])
    .nodes(d3.values(nodes))
    .links(links)
    .on('tick', tick)
    .linkDistance(100)
    .gravity(.15)
    .friction(.8)
    .linkStrength(1)
    .charge(-425)
    .chargeDistance(600)
    .start();

//LINKS
var link = svg.selectAll('.link')
    .data(links)
    .enter().append('line')
    .attr('class', 'link');

//NODES
var node = svg.selectAll('.node')
    .data(force.nodes())
    .enter().append('circle')
    .attr('class', 'node')
    .attr('r', width * 0.01)

//LABELS
var text_center = false;
var nominal_text_size = 12;
var max_text_size = 22;
var nominal_base_node_size = 8;
var max_base_node_size = 36;    
var size = d3.scale.pow().exponent(1)
  .domain([1,100])
  .range([8,24]);
    
var text = svg.selectAll(".text")
    .data(nodes)
    .enter().append("text")
    .attr("dy", ".35em")
	.style("font-size", nominal_text_size + "px")

	if (text_center)
	 text.text(function(d) { return d.name; })
	.style("text-anchor", "middle");
    
	else 
	text.attr("dx", function(d) {return (size(d.size)|| nominal_base_node_size);})
    .text(function(d) { return '\u2002'+d.name; });

//ZOOM AND PAN
function redraw() {
    svg.attr("transform",
        "translate(" + d3.event.translate + ")"
        + " scale(" + d3.event.scale + ")");
    }
    
var drag = force.drag()
      .on("dragstart", function(d) {
		d3.event.sourceEvent.stopPropagation();
      });

//NODES IN SPACE
function tick(e) {

    text.attr("transform", function(d) { return "translate(" + d.x + "," + d.y + ")"; });
    
    node.attr('cx', function(d) { return d.x; })
        .attr('cy', function(d) { return d.y; })
        .call(force.drag);

    link.attr('x1', function(d) { return d.source.x; })
        .attr('y1', function(d) { return d.source.y; })
        .attr('x2', function(d) { return d.target.x; })
        .attr('y2', function(d) { return d.target.y; });

};

//AUTOCOMPLETE SEARCH
var optArray = [];
for (var i = 0; i < nodes.length - 1; i++) {
    optArray.push(nodes[i].name);
}
optArray = optArray.sort();

//$(function () {
//    $("#search").autocomplete({
//        source: optArray
//    });
//});
window.searchNode = searchNode;

function searchNode() {
    var selectedVal = document.getElementById('search').value;
    if (selectedVal == 'none') {}
    else {
        var selected = node.filter(function (d, i) {
            return d.name != selectedVal;
        });
         var selectedText = text.filter(function (d, i) {
            return d.name != selectedVal;
        });
        selected.style("opacity", "0");
        selectedText.style("opacity", "0");
        var link = svg.selectAll(".link")
            link.style("opacity", "0");
        d3.selectAll(".node, .link, .text").transition()
            .duration(3000)
            .style("opacity", '1');
            
            var selectedNode = node
                .filter(function (d, i) { return d.name == selectedVal; })
                .datum();
                
            var scale = zoom.scale();
            var desiredPosition = { x: 200, y: 200 }; // constants, set to svg center point
            zoom.translate([desiredPosition.x - selectedNode.x*scale, desiredPosition.y - selectedNode.y*scale]);
            zoom.event(svg);
    }
}

fade.js
var svg = d3.select("svg");

var width = svg.attr("width");
var height = svg.attr("height");
var margin = {top: 10, right: 10, bottom: 30, left: 10};


var url_pre = "localhost:5000/";
const linkedByIndex = {};


svg_nodes = svg.call(d3.zoom().on("zoom", zoomed)).append("g");

svg_nodes.append("defs").append("marker")
    .attr("id", "arrow")
    .attr("viewBox", "0 -5 10 10")
    .attr("refX", 20)
    .attr("refY", 0)
    .attr("markerWidth", 8)
    .attr("markerHeight", 8)
    .attr("orient", "auto")
    .append("svg:path")
    .attr("d", "M0,-5L10,0L0,5");
svg_legend = svg.call(d3.zoom().on("zoom", zoomed)).append("g");

var color = d3.scaleOrdinal(d3.schemeCategory10);


var simulation = d3.forceSimulation()
    .force("link", d3.forceLink().id(function (d) {
        return d.id;
    }))
    .force("charge", d3.forceManyBody().strength(-200))
    .force("center", d3.forceCenter(width / 2, height / 2));

d3.json("long_list.json", createGraph);

function createGraph(error, graph) {

    if (error) throw error;


    graph.links.forEach(d => {
        linkedByIndex[`${d.source},${d.target}`] = 1;
    });

    var link = svg_nodes.append("g")
        .attr("class", "links")
        .selectAll("line")
        .data(graph.links)
        .enter().append("line")
        .attr("stroke", function (d) {
            return color(d.type + 10);
        })
        .attr("marker-end", "url(#arrow)");


    var KGPLVar = svg_nodes.append("g")
        .attr("class", "nodes")
        .selectAll("circle")
        .data(graph.var)
        .enter().append("circle")
        .attr("r", 10)
        .attr("fill", function (d) {
            if (d.root == "true") return color(d.root);
            return color(d.type);
        })
        .on('mouseover', fade(0.1))
        .on('mouseout', fade(1))
        .on("dblclick", openLink())
        .call(d3.drag()
            .on("start", dragstarted)
            .on("drag", dragged)
            .on("end", dragended));


    var KGPLVal = svg_nodes.append("g")
        .attr("class", "nodes")
        .selectAll("rect")
        .data(graph.val)
        .enter().append("rect")
        .attr("width", 20)
        .attr("height", 20)
        .attr("fill", function (d) {
            if (d.root == "true") return color(d.root);
            return color(d.type);
        })
        .on('mouseover', fade(0.1))
        .on('mouseout', fade(1))
        .on("dblclick", openLink())
        .call(d3.drag()
            .on("start", dragstarted)
            .on("drag", dragged)
            .on("end", dragended));

    const all_nodes = graph.var.concat(graph.val);
    var node_labels = svg_nodes.append("g").attr("class", "labels").selectAll("g")
        .data(all_nodes)
        .enter().append("g")
        .append("text")
        .attr("x", 14)
        .attr("y", ".31em")
        .style("font-family", "sans-serif")
        .style("font-size", "0.7em")
        .text(function (d) {
            return d.id;
        });


    KGPLVal.append("title")
        .text(function (d) {
            return d.id;
        });

    KGPLVar.append("title")
        .text(function (d) {
            return d.id;
        });
    var keys = graph.users;

    var user_legends = svg_legend.selectAll("legend_line")
        .data(keys)
        .enter().append("text")
        .attr("x", -width / 4)
        .attr("y", function (d, i) {
            return i * 10
        }) // 100 is where the first dot appears. 25 is the distance between dots
        .style("fill", function (d) {
            return color(d[1])
        })
        .text(function (d) {
            return d[0]
        })
        .style("font-size", "0.5em")

        .call(d3.drag()
            .subject(function () {
                var t = d3.select(this);
                // console.log(this.parentNode);
                return {x: t.attr('x'), y: t.attr('y')}
            })
            .on('drag', function (d, i) {
                // console.log('dragging text');
                d3.select(this).attr('y', function () {
                    return d3.mouse(this)[1];
                });
                d3.select(this).attr('x', function () {
                    return d3.mouse(this)[0];
                });

            }));


    simulation
        .nodes(all_nodes)
        .on("tick", ticked);
    console.log(graph.links);
    simulation.force("link")
        .links(graph.links);


    function ticked() {
        link
            .attr("x1", function (d) {
                return d.source.x;
            })
            .attr("y1", function (d) {
                return d.source.y;
            })
            .attr("x2", function (d) {
                return d.target.x;
            })
            .attr("y2", function (d) {
                return d.target.y;
            });

        KGPLVar
            .attr("cx", function (d) {
                return d.x;
            })
            .attr("cy", function (d) {
                return d.y;
            });
        KGPLVal
            .attr("x", function (d) {
                return d.x - 10;
            })
            .attr("y", function (d) {
                return d.y - 10;
            });
        node_labels
            .attr("transform", function (d) {
                return "translate(" + d.x + "," + d.y + ")";
            });
    }


    function fade(opacity) {
        return d => {
            KGPLVal.style('stroke-opacity', function (o) {
                console.log(isConnected(d,o));
                console.log(d);
                console.log(o);

                const thisOpacity = isConnected(d.id, o.id) ? 1 : opacity;

                this.setAttribute('fill-opacity', thisOpacity);
                return thisOpacity;
            });

            KGPLVar.style('stroke-opacity', function (o) {
                const thisOpacity = isConnected(d.id, o.id) ? 1 : opacity;
                this.setAttribute('fill-opacity', thisOpacity);
                return thisOpacity;
            });

            link.style('stroke-opacity', o => (o.source === d || o.target === d ? 1 : opacity));
            link.attr('marker-end', o => (opacity === 1 || o.source === d || o.target === d ? 'url(#arrow)' : 'url(#end-arrow-fade)'));
        };
    }
}


function dragstarted(d) {
    if (!d3.event.active) simulation.alphaTarget(0.3).restart();
    d.fx = d.x;
    d.fy = d.y;
}

function dragged(d) {
    d.fx = d3.event.x;
    d.fy = d3.event.y;
}

function dragended(d) {
    if (!d3.event.active) simulation.alphaTarget(0);
    d.fx = null;
    d.fy = null;
}

function zoomed() {
    svg_nodes.attr("transform", "translate(" + d3.event.transform.x + "," + d3.event.transform.y + ")" + " scale(" + d3.event.transform.k + ")");
    svg_legend.attr("transform", "translate(" + d3.event.transform.x + "," + d3.event.transform.y + ")" + " scale(" + d3.event.transform.k + ")");
}

function openLink() {
    return function (d) {
        var url = "";
        if (d.id != "") {
            url = url_pre.concat(d.id)
        } else {
            console.log("ERROR: No url")
        }
        window.open(url)
    }
}


function isConnected(a, b) {
    return linkedByIndex[`${a},${b}`] || linkedByIndex[`${b},${a}`] || a === b;
}


/*
var data = {
    "val": [
        {
            "id": "val/0",
            "type": 1,
            "creater":"UserA"
        },
        {
            "id": "val/1",
            "type": 1,
            "creater":"UserA"
        }],
    "var": [{
        "id": "var/0",
        "type": 2,
        "creater":"UserB"
    }],
    "links": [
        {
            "source": "var/0",
            "target": "val/1",
            "type": 1
        }]
};*/

// createGraph(false, data);



hyperlinked.js
// http://jsfiddle.net/Bull/4btFx/1/
var data = {
    "nodes": [{
        "name": "Action 4",
        "type": 5,
        "slug": "",
        "value": 265000
    }, {
        "name": "Action 5",
        "type": 6,
        "slug": "",
        "value": 23000
    }, {
        "name": "Action 3",
        "type": 4,
        "slug": "",
        "value": 115000
    }, {
        "name": "Yahoo",
        "type": 1,
        "slug": "www.yahoo.com",
        "entity": "company"
    }, {
        "name": "Google",
        "type": 1,
        "slug": "www.google.com",
        "entity": "company"
    }, {
        "name": "Action 1",
        "type": 2,
        "slug": "",
    }, {
        "name": "Action 2",
        "type": 3,
        "slug": "",
    }, {
        "name": "Bing",
        "type": 1,
        "slug": "www.bing.com",
        "entity": "company"
    }, {
        "name": "Yandex",
        "type": 1,
        "slug": "www.yandex.com)",
        "entity": "company"
    }],
        "links": [{
        "source": 0,
        "target": 3,
        "value": 10
    }, {
        "source": 4,
        "target": 3,
        "value": 1
    }, {
        "source": 1,
        "target": 7,
        "value": 10
    }, {
        "source": 2,
        "target": 4,
        "value": 10
    }, {
        "source": 4,
        "target": 7,
        "value": 1
    }, {
        "source": 4,
        "target": 5,
        "value": 10
    }, {
        "source": 4,
        "target": 6,
        "value": 10
    }, {
        "source": 8,
        "target": 4,
        "value": 1
    }]
}



var w = $(document).width(),
    h = $(document).height(),
    radius = d3.scale.log().domain([0, 312000]).range(["10", "50"]);

var vis = d3.select("body").append("svg:svg")
    .attr("width", w)
    .attr("height", h);

vis.append("defs").append("marker")
    .attr("id", "arrowhead")
    .attr("refX", 17 + 3) /*must be smarter way to calculate shift*/
.attr("refY", 2)
    .attr("markerWidth", 6)
    .attr("markerHeight", 4)
    .attr("orient", "auto")
    .append("path")
    .attr("d", "M 0,0 V 4 L6,2 Z"); //this is actual shape for arrowhead

//d3.json(data, function(json) {
var force = self.force = d3.layout.force()
    .nodes(data.nodes)
    .links(data.links)
    .distance(100)
    .charge(-1000)
    .size([w, h])
    .start();



var link = vis.selectAll("line.link")
    .data(data.links)
    .enter().append("svg:line")
    .attr("class", function (d) {
    return "link" + d.value + "";
})
    .attr("x1", function (d) {
    return d.source.x;
})
    .attr("y1", function (d) {
    return d.source.y;
})
    .attr("x2", function (d) {
    return d.target.x;
})
    .attr("y2", function (d) {
    return d.target.y;
})
    .attr("marker-end", function (d) {
    if (d.value == 1) {
        return "url(#arrowhead)"
    } else {
        return " "
    };
});


function openLink() {
    return function (d) {
        var url = "";
        if (d.slug != "") {
            url = d.slug
        } //else if(d.type == 2) {
        //url = "clients/" + d.slug
        //} else if(d.type == 3) {
        //url = "agencies/" + d.slug
        //}
        window.open("//" + url)
    }
}




var node = vis.selectAll("g.node")
    .data(data.nodes)
    .enter().append("svg:g")
    .attr("class", function (d) {
        if (d.entity === "company") {
           return "company node";
        } else {
           return "circle node";
        }
    })
    .call(force.drag);

d3.selectAll(".company").append("rect")
    .attr("width", 20)
    .attr("height", 20)
    .attr("class", function (d) {
    return "node type" + d.type
});

d3.selectAll(".circle").append("circle")
    .attr("class", function (d) {
    return "node type" + d.type
})
    .attr("r", function (d) {
    return radius(d.value) || 10
})
//.style("fill", function(d) { return fill(d.type); })
.call(force.drag);

node.append("svg:image")
    .attr("class", "circle")
    .attr("xlink:href", function (d) {
    return d.img_href
})
    .attr("x", "-16px")
    .attr("y", "-16px")
    .attr("width", "32px")
    .attr("height", "32px")
    .on("click", openLink());

node.append("svg:text")
    .attr("class", "nodetext")
    .attr("dx", 0)
    .attr("dy", ".35em")
    .attr("text-anchor", "middle")
    .text(function (d) {
    return d.name
});

force.on("tick", function () {
    link.attr("x1", function (d) {
        return d.source.x;
    })
        .attr("y1", function (d) {
        return d.source.y;
    })
        .attr("x2", function (d) {
        return d.target.x;
    })
        .attr("y2", function (d) {
        return d.target.y;
    });

    node.attr("transform", function (d) {
        return "translate(" + d.x + "," + d.y + ")";
    });
});
//});

zoom_detail.js
var svg = d3.select("svg");

var width = svg.attr("width");
var height = svg.attr("height");
// var margin = {top: 10, right: 10, bottom: 30, left: 10};


var url_pre = "localhost:5000/";
const linkedByIndex = {};


svg_nodes = svg.call(d3.zoom().on("zoom", zoomed)).append("g");

svg_nodes.append("defs").append("marker")
    .attr("id", "arrow")
    .attr("viewBox", "0 -5 10 10")
    .attr("refX", 20)
    .attr("refY", 0)
    .attr("markerWidth", 8)
    .attr("markerHeight", 8)
    .attr("orient", "auto")
    .append("svg:path")
    .attr("d", "M0,-5L10,0L0,5");
svg_legend = svg.call(d3.zoom().on("zoom", zoomed)).append("g");

var color = d3.scaleOrdinal(d3.schemeCategory10);


var simulation = d3.forceSimulation()
    .force("link", d3.forceLink().id(function (d) {
        return d.id;
    }))
    .force("charge", d3.forceManyBody().strength(-200))
    .force("center", d3.forceCenter(width / 2, height / 2));

d3.json("zoom_detail.json", createGraph);
var zoom = d3.zoom().on("zoom", zoomed);

/*

function redraw() {
    svg.attr("transform",
        "translate(" + d3.event.translate + ")"
        + " scale(" + d3.event.scale + ")");
}
*/

function createGraph(error, graph) {

    if (error) throw error;

    pack = data => d3.pack()
        .size([width, height])
        .padding(3)
        (d3.hierarchy(data)
            .sum(d => d.value)
            .sort((a, b) => b.value - a.value));


    graph.links.forEach(d => {
        linkedByIndex[`${d.source},${d.target}`] = 1;
    });

    var link = svg_nodes.append("g")
        .attr("class", "links")
        .selectAll("line")
        .data(graph.links)
        .enter().append("line")
        .attr("stroke", function (d) {
            return color(d.type + 10);
        })
        .attr("marker-end", "url(#arrow)");


    var KGPLVar = svg_nodes.append("g")
        .attr("class", "nodes")
        .selectAll("circle")
        .data(graph.var)
        .enter().append("circle")
        .attr("r", 10)
        .attr("fill", function (d) {
            if (d.root == "true") return color(d.root);
            return color(d.type);
        })
        .on('mouseover', fade(0.1))
        .on('mouseout', fade(1))
        .on("dblclick", openLink())
        .call(d3.drag()
            .on("start", dragstarted)
            .on("drag", dragged)
            .on("end", dragended));


    var KGPLVal = svg_nodes.append("g")
        .attr("class", "nodes")
        .selectAll("rect")
        .data(graph.val)
        .enter().append("rect")
        .attr("width", 20)
        .attr("height", 20)
        .attr("fill", function (d) {
            if (d.root == "true") return color(d.root);
            return color(d.type);
        })
        .on('mouseover', fade(0.1))
        .on('mouseout', fade(1))
        .on("dblclick", openLink())
        .call(d3.drag()
            .on("start", dragstarted)
            .on("drag", dragged)
            .on("end", dragended));

    const all_nodes = graph.var.concat(graph.val);
    var node_labels = svg_nodes.append("g").attr("class", "labels").selectAll("g")
        .data(all_nodes)
        .enter().append("g")
        .append("text")
        .attr("x", 14)
        .attr("y", ".31em")
        // .attr("text-anchor", "middle")
        .style("font-family", "sans-serif")
        .style("font-size", "0.7em")
        .text(function (d) {
            return d.id;
        });


    KGPLVal.append("title")
        .text(function (d) {
            return d.id;
        });

    KGPLVar.append("title")
        .text(function (d) {
            return d.id;
        });
    var keys = graph.users;

    var user_legends = svg_legend.selectAll("legend_line")
        .data(keys)
        .enter().append("text")
        .attr("x", 0)
        .attr("y", function (d, i) {
            return i * 15
        }) // 100 is where the first dot appears. 25 is the distance between dots
        .style("fill", function (d) {
            return color(d[1])
        })
        .text(function (d) {
            return d[0]
        })
        .style("font-size", "0.7em")

        .call(d3.drag()
            .subject(function () {
                var t = d3.select(this);
                // console.log(this.parentNode);
                return {x: t.attr('x'), y: t.attr('y')}
            })
            .on('drag', function (d, i) {
                // console.log('dragging text');
                d3.select(this).attr('y', function () {
                    return d3.mouse(this)[1];
                });
                d3.select(this).attr('x', function () {
                    return d3.mouse(this)[0];
                });

            }));


    simulation
        .nodes(all_nodes)
        .on("tick", ticked);
    // console.log(graph.links);
    simulation.force("link")
        .links(graph.links);


    function ticked() {
        link
            .attr("x1", function (d) {
                return d.source.x;
            })
            .attr("y1", function (d) {
                return d.source.y;
            })
            .attr("x2", function (d) {
                return d.target.x;
            })
            .attr("y2", function (d) {
                return d.target.y;
            });

        KGPLVar
            .attr("cx", function (d) {
                return d.x;
            })
            .attr("cy", function (d) {
                return d.y;
            });
        KGPLVal
            .attr("x", function (d) {
                return d.x - 10;
            })
            .attr("y", function (d) {
                return d.y - 10;
            });
        node_labels
            .attr("transform", function (d) {
                return "translate(" + d.x + "," + d.y + ")";
            });
    }


    function fade(opacity) {
        return d => {
            KGPLVal.style('stroke-opacity', function (o) {
                /*                console.log(isConnected(d, o));
                                console.log(d);
                                console.log(o);*/

                const thisOpacity = isConnected(d.id, o.id) ? 1 : opacity;

                this.setAttribute('fill-opacity', thisOpacity);
                return thisOpacity;
            });

            KGPLVar.style('stroke-opacity', function (o) {
                const thisOpacity = isConnected(d.id, o.id) ? 1 : opacity;
                this.setAttribute('fill-opacity', thisOpacity);
                return thisOpacity;
            });

            link.style('stroke-opacity', o => (o.source === d || o.target === d ? 1 : opacity));
            link.attr('marker-end', o => (opacity === 1 || o.source === d || o.target === d ? 'url(#arrow)' : 'url(#end-arrow-fade)'));
        };
    }


}


function dragstarted(d) {
    if (!d3.event.active) simulation.alphaTarget(0.3).restart();
    d.fx = d.x;
    d.fy = d.y;
}

function dragged(d) {
    d.fx = d3.event.x;
    d.fy = d3.event.y;
}

function dragended(d) {
    if (!d3.event.active) simulation.alphaTarget(0);
    d.fx = null;
    d.fy = null;
}

function zoomed() {
    svg_nodes.attr("transform", "translate(" + d3.event.transform.x + "," + d3.event.transform.y + ")" + " scale(" + d3.event.transform.k + ")");
    // svg_legend.attr("transform", "translate(" + d3.event.transform.x + "," + d3.event.transform.y + ")" + " scale(" + d3.event.transform.k + ")");
}

function openLink() {
    return function (d) {
        var url = "";
        if (d.id != "") {
            url = url_pre.concat(d.id)
        } else {
            console.log("ERROR: No url")
        }
        window.open(url)
    }
}


function isConnected(a, b) {
    return linkedByIndex[`${a},${b}`] || linkedByIndex[`${b},${a}`] || a === b;
}


function searchNode() {
    var selectedVal = document.getElementById('search').value;
    if (selectedVal == 'none') {
    } else {
        var kgpl_var = svg.selectAll('circle'); //KGPLVar
        var kgpl_val = svg.selectAll("rect"); //KGPLVal
        // var text = svg.selectAll("node");
        // console.log(text)
        var selected_var = kgpl_var.filter(function (d, i) {
            // console.log(d.id)
            return d.id === selectedVal;
        });
        var selected_val = kgpl_val.filter(function (d, i) {
            // console.log(d.id)
            return d.id === selectedVal;
        });
        // console.log(selected_val);
        // console.log(selected_var);

        if (selected_val.size() !== 0) {
            selectedNode = selected_val.datum();
        } else if (selected_var.size() !== 0) {
            selectedNode = selected_var.datum();
        } else {
            console.log("Node not existed!");
            return;
        }
        // var selectedText = text.filter(function (d, i) {
        //     return d.name !== selectedVal;
        // });
        // selected.style("opacity", "0");
        // selectedText.style("opacity", "0");
        // var link = svg.selectAll("link")
        /*        link.style("opacity", "0");
                d3.selectAll("link, circle, rect").transition()
                    .duration(3000)
                    .style("opacity", '1');*/


        /*        var selectedNode = node
                    .filter(function (d, i) {
                        return d.name === selectedVal;
                    })
                    .datum();*/

        // var scale = zoom.scale();
        var desiredPosition = {
            x: width / 2 - selectedNode.x,
            y: height / 2 - selectedNode.y
        }; // constants, set to svg center point
        // zoom.translate([desiredPosition.x - selectedNode.x * scale, desiredPosition.y - selectedNode.y * scale]);
        var scale = 1;
        svg.transition()
            .duration(750)
            // .call(zoom.translate(translate).scale(scale).event); // not in d3 v4
            .call(zoom.transform, d3.zoomIdentity.translate(desiredPosition["x"], desiredPosition["y"]).scale(scale)); // updated for d3 v4


        // zoom.translate([desiredPosition.x - selectedNode.x, desiredPosition.y - selectedNode.y]);

        // zoom.event(svg);
    }
}


/*
var data = {
    "val": [
        {
            "id": "val/0",
            "type": 1,
            "creater":"UserA"
        },
        {
            "id": "val/1",
            "type": 1,
            "creater":"UserA"
        }],
    "var": [{
        "id": "var/0",
        "type": 2,
        "creater":"UserB"
    }],
    "links": [
        {
            "source": "var/0",
            "target": "val/1",
            "type": 1
        }]
};*/

// createGraph(false, data);

